## Q1. Create a new notebook for this task. Be sure to import the relevant libraries, along with your ords_prods_merge dataframe, which should include your newly derived columns from the previous Exercise.

In [4]:
#Importing Libraries
import pandas as pd
import numpy as np
import os

## Creating path to main project 'InstacartBasket'

In [5]:
path = r"C:\Users\Leeza\Desktop\Data Analytics\Data Immersion\Python\Instacart Basket Analysis"

In [6]:
# Importing orders_products_merged_updated file
ords_prods_merge=pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_updated.pkl'))

## Q2. In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [14]:
# performing aggregation on order_number group by department_id
ords_prods_merge.groupby(['department_id'])['order_number'].mean()

department_id
1     15.457838
2     17.277920
3     17.170395
4     17.811403
5     15.215751
6     16.439806
7     17.225802
8     15.340650
9     15.895474
10    20.197148
11    16.170638
12    15.887671
13    16.583536
14    16.773669
15    16.165037
16    17.665606
17    15.694469
18    19.310397
19    17.177343
20    16.473447
21    22.902379
Name: order_number, dtype: float64

## Q3. Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

### Answer:
The subset result focuses on a specific set of department IDs, so it provides insights into the order numbers within that particular group.
The entire DataFrame result includes all department IDs, offering a broader perspective on the overall order numbers.
The average order numbers can vary significantly between different department IDs, and this variation becomes more apparent when looking at subsets.

## Q4. Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [15]:
# create new column to store max order to identify loyalty customers
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [16]:
ords_prods_merge.head(15)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      0.0         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   

   product_name  aisle_id  department_id  prices    price_range_loc  \
0          Soda        77              7     9.0  Mid-range product   
1          Soda        77              7     9.0  Mid-range product   
2          Soda        77              7     9.0  Mid-range product   
3          Soda        77              7     9.0  Mid-range product   
4          Soda        77              7     9.0  Mid-range product   
5          Soda        77              7     9.0  Mid-range product   
6          Soda        77              7     9.0  Mid-range product   
7          Soda        77              7     9.0  Mid-range product   
8          Soda        77              7     9.0  Mid-range product   
9          Soda        77              7     9.0  Mid-range product   
10         Soda        77              7     9.0  Mid-range product   
11         Soda        77              7     9.0  Mid-range product   
12         Soda        77              7     9.0  Mid-range product   
13         Soda        77              7     9.0  Mid-range product   
14         Soda        77              7     9.0  Mid-range product   

       Busiest_day        Busiest_days busiest_period_of_day  max_order  
0   Regularly busy      Regularly busy        Average Orders         10  
1   Regularly busy   Second Least busy        Average Orders         10  
2   Regularly busy   Second Least busy           Most Orders         10  
3       Least busy          Least busy        Average Orders         10  
4       Least busy          Least busy           Most Orders         10  
5   Regularly busy   

In [17]:
# setting loyalty flag as 'loyal customer'
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [18]:
# setting loyalty flag as 'regular customer'
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [19]:
# setting loyalty flag as 'new customer'
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [20]:
# frequency of your new “loyalty_flag” column
ords_prods_merge['loyalty_flag'].value_counts()

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [21]:
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(10)

user_id  loyalty_flag  order_number
0        1  New customer             1
1        1  New customer             2
2        1  New customer             3
3        1  New customer             4
4        1  New customer             5
5        1  New customer             6
6        1  New customer             7
7        1  New customer             8
8        1  New customer             9
9        1  New customer            10

## Q5. The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [26]:
# aggregation function to get price statistics
ords_prods_merge.groupby(['loyalty_flag']).agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386335  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

## Answer:
Mean Price (Average Price):

Loyal customer: The mean price is approximately 10.39.
New customer: The mean price is approximately 13.29.
Regular customer: The mean price is approximately 12.50.

Minimum Price:

For all loyalty categories, the minimum price is 1.0. This indicates that the lowest-priced products are the same for all three groups.
Maximum Price:

For all loyalty categories, the maximum price is 99999.0. This suggests that the highest-priced products are the same for all three groups.

## Q6. The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria: If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.” If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [30]:
# create new column to store average of price to identify types of spenders
ords_prods_merge['avg_prices'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [31]:
ords_prods_merge.head(15)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      0.0         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   

   product_name  aisle_id  department_id  prices    price_range_loc  \
0          Soda        77              7     9.0  Mid-range product   
1          Soda        77              7     9.0  Mid-range product   
2          Soda        77              7     9.0  Mid-range product   
3          Soda        77              7     9.0  Mid-range product   
4          Soda        77              7     9.0  Mid-range product   
5          Soda        77              7     9.0  Mid-range product   
6          Soda        77              7     9.0  Mid-range product   
7          Soda        77              7     9.0  Mid-range product   
8          Soda        77              7     9.0  Mid-range product   
9          Soda        77              7     9.0  Mid-range product   
10         Soda        77              7     9.0  Mid-range product   
11         Soda        77              7     9.0  Mid-range product   
12         Soda        77              7     9.0  Mid-range product   
13         Soda        77              7     9.0  Mid-range product   
14         Soda        77              7     9.0  Mid-range product   

       Busiest_day        Busiest_days busiest_period_of_day  max_order  \
0   Regularly busy      Regularly busy        Average Orders         10   
1   Regularly busy   Second Least busy        Average Orders         10   
2   Regularly busy   Second Least busy           Most Orders         10   
3       Least busy          Least busy        Average Orders         10   
4       Least busy          Least busy           Most Orders         10   
5   Regularly b

In [38]:
# setting spending flag as 'low spender'
ords_prods_merge.loc[ords_prods_merge['avg_prices'] <10, 'spending_flag'] = 'Low Spender'

In [39]:
# setting spending flag as 'High spender'
ords_prods_merge.loc[ords_prods_merge['avg_prices'] >=10, 'spending_flag'] = 'High Spender'

In [40]:
ords_prods_merge['spending_flag'].value_counts()

Low Spender     31770614
High Spender      634245
Name: spending_flag, dtype: int64

In [41]:
ords_prods_merge[['user_id', 'spending_flag', 'avg_prices']].head(10)

user_id spending_flag  avg_prices
0        1   Low Spender    6.367797
1        1   Low Spender    6.367797
2        1   Low Spender    6.367797
3        1   Low Spender    6.367797
4        1   Low Spender    6.367797
5        1   Low Spender    6.367797
6        1   Low Spender    6.367797
7        1   Low Spender    6.367797
8        1   Low Spender    6.367797
9        1   Low Spender    6.367797

## Q7. In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows: If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.” If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.” If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [36]:
# create new column to store median of 'days_since_prior_order' to identify frequent customer
ords_prods_merge['median_dspo'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [37]:
ords_prods_merge.head(10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   
5   3367565        1             6                   2                  7   
6    550135        1             7                   1                  9   
7   3108588        1             8                   1                 14   
8   2295261        1             9                   1                 16   
9   2550362        1            10                   4                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     0.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   
5                    19.0         196                  1          1   
6                    20.0         196                  1          1   
7                    14.0         196                  2          1   
8                     0.0         196                  4          1   
9                    30.0         196                  1          1   

  product_name  ...  prices    price_range_loc     Busiest_day  \
0         Soda  ...     9.0  Mid-range product  Regularly busy   
1         Soda  ...     9.0  Mid-range product  Regularly busy   
2         Soda  ...     9.0  Mid-range product  Regularly busy   
3         Soda  ...     9.0  Mid-range product      Least busy   
4         Soda  ...     9.0  Mid-range product      Least busy   
5         Soda  ...     9.0  Mid-range product  Regularly busy   
6         Soda  ...     9.0  Mid-range product  Regularly busy   
7         Soda  ...     9.0  Mid-range product  Regularly busy   
8         Soda  ...     9.0  Mid-range product  Regularly busy   
9         Soda  ...     9.0  Mid-range product      Least busy   

         Busiest_days busiest_period_of_day max_order  loyalty_flag  \
0      Regularly busy        Average Orders        10  New customer   
1   Second Least busy        Average Orders        10  New customer   
2   Second Least busy           Most Orders        10  New customer   
3          Least busy        Average Orders        10  New customer   
4          Least busy           Most Orders        10  New customer   
5      Regularly busy        Average Orders        10  New customer   
6  Second Busiest day           Most Orders        10  New customer   
7  Second Busiest day           Most Orders        10  New customer   
8  Second Busiest day           Most Orders        10  New customer   
9          Least busy        Average Orders        10  New customer   

   avg_prices spending_flag  median_dspo  
0    6.367797   Low Spender         20.0  
1    6.367797   Low Spender         20.0  
2    6.367797   Low Spender         20.0  
3    6.367797   Low Spender         20.0  
4    6.367797   Low Spender         20.0  
5    6.367797   Low Spender         20.0  
6    6.367797   Low Spender         20.0  
7    6.367797   Low Spender         20.0  
8    6.367797   Low Spender         20.0  
9    6.367797   Low Spender         20.0  

[10 rows x 22 columns]

In [42]:
# setting order_frequency flag as 'Non-frequent customer'
ords_prods_merge.loc[ords_prods_merge['median_dspo'] >20, 'order_frequency'] = 'Non-frequent customer'

In [43]:
# setting order_frequency flag as 'Regular customer'
ords_prods_merge.loc[(ords_prods_merge['median_dspo'] >10) & (ords_prods_merge['median_dspo'] < 20), 'order_frequency'] = 'Regular customer'

In [44]:
# setting order_frequency flag as 'Frequent customer'
ords_prods_merge.loc[ords_prods_merge['median_dspo'] <= 10, 'order_frequency'] = 'Frequent customer'

In [45]:
ords_prods_merge['order_frequency'].value_counts()

Frequent customer        22796659
Regular customer          6595192
Non-frequent customer     2686728
Name: order_frequency, dtype: int64

In [47]:
ords_prods_merge[['user_id', 'order_frequency', 'median_dspo']].head(10)

user_id order_frequency  median_dspo
0        1             NaN         20.0
1        1             NaN         20.0
2        1             NaN         20.0
3        1             NaN         20.0
4        1             NaN         20.0
5        1             NaN         20.0
6        1             NaN         20.0
7        1             NaN         20.0
8        1             NaN         20.0
9        1             NaN         20.0

## Q9. Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder.

In [48]:
# Export the new file
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_aggregated.pkl'))